src/kafunk/Consumer.fs

```diff
@@ -897,6 +911,70 @@ module Consumer =
           commitCurrentOffsetsOnGroupJoinProc)
         |> Async.Ignore }
 
+
+  /// Starts consumption from the start offset in the given range
+  /// Will stop consuming for each partition once it reaches the max boundary offset
+  /// Will stop and return the messages WITHIN the range if it overconsumes
+  let consumeRange (c:Consumer) (range:OffsetRange) : Async<seq<Message>> = 
+    async { 
+      let getMsgFromRg (msgs : seq<Partition * Message [] * Map<Partition, Offset>>) range =
+        msgs
+          |> Seq.map (fun (p, ms, _) -> p, ms)
+          |> Seq.groupBy (fst)
+          |> Seq.map (fun (p, u) -> 
+                let num = (range.endOS.[p]) - (range.startOS.[p])
+                u
+                |> Seq.map (snd)
```

Comment:
The previous `map`, `groupBy`, `map`, and `map` effectively loop over  the array 4 times. This can be done with a single `fold`.